From ce8766d359263571a748e3419874a96bb9eb6d99 Mon Sep 17 00:00:00 2001
From: "[Partner] Sergey Nizovtsev" <sn@tempesta-tech.com>
Date: Sat, 25 Jun 2022 08:50:05 +0300
Subject: [PATCH] GCORE: fpu: disable bottom-halves in FPU context

Signed-off-by: Sergey Nizovtsev <sn@tempesta-tech.com>
---
 arch/x86/kernel/fpu/core.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/arch/x86/kernel/fpu/core.c b/arch/x86/kernel/fpu/core.c
index 98e6c29e17a48..de80a1bfe6bb6 100644
--- a/arch/x86/kernel/fpu/core.c
+++ b/arch/x86/kernel/fpu/core.c
@@ -425,6 +425,13 @@ EXPORT_SYMBOL_GPL(fpu_copy_uabi_to_guest_fpstate);
 
 void kernel_fpu_begin_mask(unsigned int kfpu_mask)
 {
+	/*
+	 * We don't know in which context the function is called, but we know
+	 * preciseely that softirq uses FPU, so we have to disable softirq as
+	 * well as task preemption.
+	 */
+	local_bh_disable();
+
 	preempt_disable();
 
 	WARN_ON_FPU(!irq_fpu_usable());
@@ -454,6 +461,7 @@ void kernel_fpu_end(void)
 
 	this_cpu_write(in_kernel_fpu, false);
 	preempt_enable();
+	local_bh_enable();
 }
 EXPORT_SYMBOL_GPL(kernel_fpu_end);
 
-- 
2.38.0

